### PR TITLE
Appliance shared TLS certificate

### LINKED ISSUE #881: OVA shared TLS certificate
@andrewtchin commented on [Wed Jul 05 2017](https://github.com/vmware/vic/issues/5640)

**User Statement:**

As a user I want to be able to use the same TLS certificate across all components of the OVA, whether it is generated by the OVA or entered during the OVA deploy.

**Details:**
We currently generate a separate self signed cert for each component or require the user to enter multiple certificates (or the same certificate multiple times) in the deploy. Change this to be shared so the user can enter it once.
Also examine how we handle a user provided CA (is it added to the root store?) and how that impacts appliance initialization (skip thumbprint verification on Getting Started Page if we trust the cert on vCenter - this investigation may lead to opening another issue) 

**Acceptance Criteria:**
- [x] Share OVA generated or user provided certificate between all OVA components
- [x] Show the certificate fingerprint on the DCUI
- [x] Certificate should use the hostname if present (See #1006)

### LINKED ISSUE #1007: Update doc link in output of `systemctl status harbor/admiral/fileserver`
@stuclem commented on [Wed Apr 05 2017](https://github.com/vmware/vic/issues/4581)

**VIC version:**

OVA installer: `vic-v1.1.0-rc2-81820686.ova`

**Deployment details:**

N/A

**Steps to reproduce:**

SSH into the appliance VM and run  `systemctl status harbor/admiral/fileserver`

**Actual behavior:**
Output inludes links to the docs.

Registry: `Docs: http://github.com/vmware/harbor`
Portal: `Docs: https://github.com/vmware/admiral`
Engine: `Docs: https://github.com/vmware/vic`

**Expected behavior:**

Now that I am unifying the docs, it would be better if the link were to https://vmware.github.io/vic-product/index.html#getting-started in all cases.

### PR DESCRIPTION
**Pull request by andrewtchin**
Fixes #881 
Fixes #1007 
Towards #1179 

tested with
generated self signed DHCP
generated self signed static
custom cert DHCP
custom cert static

### CODE REVIEW

**Comment by andrewtchin** [comment]
yes i will definitely be making sure everything on the appliance uses this. i'm waiting until your systemd refactor merges first :-)

**Comment by andrewtchin** [comment]
account for encrypted private key
```Jan 09 17:06:51 server.local start_fileserver.sh[2137]: time="2018-01-09T17:06:51Z" level=info msg="Loading certificate /opt/vmware/fileserver/cert/server.crt and key /opt/vmware/fileserver/cert/server.key"
Jan 09 17:06:51 server.local start_fileserver.sh[2137]: time="2018-01-09T17:06:51Z" level=fatal msg="Failed to load certificate /opt/vmware/fileserver/cert/server.crt and key /opt/vmware/fileserver/cert/server.key: tls: failed to pars
e private key"```

**Comment by andrewtchin** [comment]
make sure this survives rebase https://github.com/vmware/vic-product/pull/1381

**Comment by andrewtchin** [comment]
screen shot 2018-02-02 at 3 42 22 pm

**Comment by andrewtchin** [comment]
@morris-jason ready for another review

File: installer/ovatools/vic-ova-ui/main.go [hunk 1]
```diff
@@ -197,3 +204,23 @@ func getMemory() string {
 	}
 	return fmt.Sprintf("%s Memory", humanize.IBytes(uint64(si.Totalram)))
 }
+
+func getTLSCertFingerprint() string {
+	certFile := "/storage/data/certs/server.crt"
+	if _, err := os.Stat(certFile); err == nil {
```
Comment (morris-jason) [inline installer/ovatools/vic-ova-ui/main.go]: Could this be simplified to:
```golang
func getTLSCertFingerprint() string {
    certFile := "/storage/data/certs/server.crt"
    certPEM, e := ioutil.ReadFile(certFile)
    if e != nil {
        log.Debugf("Read error: %s", e.Error())
        return certNotAvail
    }
    block, _ := pem.Decode([]byte(certPEM))
    if block == nil {
        log.Debugf("Parse error")
        return certNotAvail
    }
    cert, err := x509.ParseCertificate(block.Bytes)
    if err != nil {
        log.Debugf("Parse error: %s", err.Error())
        return certNotAvail
    }
return fmt.Sprintf("% X", sha1.Sum(cert.Raw))
```

File: installer/ovatools/vic-ova-ui/main.go [hunk 2]
```diff
@@ -197,3 +204,23 @@ func getMemory() string {
...
+		}
+		block, _ := pem.Decode([]byte(certPEM))
+		if block == nil {
+			return fmt.Sprintf("Parse error")
+		}
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return fmt.Sprintf("Parse error: %s", err.Error())
+		}
+		return fmt.Sprintf("% X", sha1.Sum(cert.Raw))
```
Comment (morris-jason) [inline installer/ovatools/vic-ova-ui/main.go]: Should this be "%X" instead of "% X"? Does the extra space break this?

Comment (andrewtchin) [reply installer/ovatools/vic-ova-ui/main.go]: space adds a space after each octet for readability

**Pull request merged**